pgrx-pg-sys/build/clang.rs

```diff
@@ -0,0 +1,129 @@
+/// but the way rust-bindgen does it breaks on Postgres 16 due to code like
+/// ```c
+/// #include <emmintrin.h>
+/// ```
+/// This will pull in builtin headers, but rust-bindgen uses a $CLANG_PATH lookup from clang-sys
+/// which is not guaranteed to find the clang that uses the $LIBCLANG_PATH that bindgen intends.
+///
+/// Returns the set of paths to include.
+pub(crate) fn detect_include_paths_for(
+    preferred_clang: Option<&std::path::Path>,
+) -> (bool, Vec<PathBuf>) {
+    if target_env_tracked("PGRX_BINDGEN_NO_DETECT_INCLUDES").is_some() {
+        return (false, vec![]);
+    }
+
+    // By asking bindgen for the version, we force it to pull an appropriate libclang,
+    // allowing users to override it however they would usually override bindgen.
+    let clang_major = match bindgen::clang_version() {
+        ClangVersion { parsed: Some((major, _)), full } => {
+            eprintln!("Bindgen found {full}");
+            major
+        }
+        ClangVersion { full, .. } => {
+            // If bindgen doesn't know what version it has, bail and hope for the best.
+            eprintln!("Bindgen failed to parse clang version: {full}");
+            return (true, vec![]);
+        }
+    };
+
+    // If Postgres is configured --with-llvm, then it may have recorded a CLANG to use
+    // Ask if there's a clang at the path that Postgres would use for JIT purposes.
+    // Unfortunately, the responses from clang-sys include clangs from far-off paths,
+    // so we can only use clangs that match bindgen's libclang major version.
+    if let Some(ClangSys { path, version: Some(v), c_search_paths, .. }) =
+        ClangSys::find(preferred_clang, &[])
+    {
+        if Some(&*path) == preferred_clang && v.Major as u32 == clang_major {
+            return (false, c_search_paths.unwrap_or_default());
+        }
+    }
+
+    // Oh no, still here?
+    // Let's go behind bindgen's back to get libclang's path
+    let libclang_path =
+        clang_sys::get_library().expect("libclang should have been loaded?").path().to_owned();
+    eprintln!("found libclang at {}", libclang_path.display());
+    // libclang will probably be in a dynamic library directory,
+    // which means it will probably be adjacent to its headers, e.g.
+    // - "/usr/lib/libclang-${CLANG_MAJOR}.so.${CLANG_MAJOR}.${CLANG_MINOR}"
+    // - "/usr/lib/clang/${CLANG_MAJOR}/include"
+    let clang_major_fmt = clang_major.to_string();
+    let mut paths = vec![];
+    // by adjacent, that does not mean it is always immediately so, e.g.
+    // - "/usr/lib/x86_64-linux-gnu/libclang-${CLANG_MAJOR}.so.${CLANG_MAJOR}.${CLANG_MINOR}.${CLANG_SUBMINOR}"
+    // - "/usr/lib/clang/${CLANG_MAJOR}/include"
+    // or
+    // - "/usr/lib64/libclang-${CLANG_MAJOR}.so.${CLANG_MAJOR}.${CLANG_MINOR}.${CLANG_SUBMINOR}"
+    // - "/usr/lib/clang/${CLANG_MAJOR}/include"
+    // so, crawl back up the ancestral tree
+    for ancestor in libclang_path.ancestors() {
```

Comment:
This is the part where the secret ingredient is Crime. The idea is that this crime starts its journey relative to the libclang.so dir so that we ideally find the right headers to pull even if we're starting from an unusual place.